### DISCUSSION TITLE
CAN Auslesen bricht häufig ab

### DISCUSSION THREAD

**Discussion by timogru** [open3e/open3e#67] [General]
Hallo,

nach dem mein System grundsätzlich funktioniert und ich Daten aus meiner VX3 auslesen kann, bricht nun die Datenübertragung häufig nach ein paar Tagen ab. In der nohup.out habe ich z.B. diesen Eintrag dazu

```
2024-01-26 08:53:19 [ERROR] UdsClient: [TimeoutException] : Did not receive response in time. Global request timeout time has expired (timeout=20.000 sec)
Traceback (most recent call last):
  File "/home/raspberry/open3e/Open3Eclient.py", line 362, in <module>
    val,idstr = dicEcus[ecudid[0]].readByDid(ecudid[1], args.raw)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Es gibt aber auch andere Fehler wie z.B.:

```
2024-01-20 12:58:34 [ERROR] UdsClient: [UnexpectedResponseException] : ReadDataByIdentifier service execution returned a valid response but unexpected. Details : Server returned values for 1 data >
Traceback (most recent call last):
  File "/home/raspberry/open3e/Open3Eclient.py", line 362, in <module>
    val,idstr = dicEcus[ecudid[0]].readByDid(ecudid[1], args.raw)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/raspberry/open3e/Open3Eclass.py", line 114, in readByDid
    response = self.uds_client.read_data_by_identifier([did])
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/raspberry/.local/lib/python3.11/site-packages/udsoncan/client.py", line 174, in decorated
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/raspberry/.local/lib/python3.11/site-packages/udsoncan/client.py", line 475, in read_data_by_identifier
    raise UnexpectedResponseException(
udsoncan.exceptions.UnexpectedResponseException: ReadDataByIdentifier service execution returned a valid response but unexpected. Details : Server returned values for 1 data identifier that were n>
```

Was könnte ich machen, damit das stabil läuft?

VG, Timo

**Comment by philippoo66** [comment]
ist das mit der neuen Version, wo die Kommunikation schon auf python-can umgestellt ist?

**Comment by MyHomeMyData** [comment]
@timogru Wann und von welchem Branch hast Du open3e installiert?

**Comment by timogru** [comment]
Hi @MyHomeMyData und @philippoo66 
Freitags eine Frage stellen und dann das ganze Wochenende Kinderprogramm zu haben, war jetzt nicht die beste Idee, sorry.

Das ist wohl noch die alte Version. Denn ich hatte mit: `pip3 uninstall can-isotp && pip3 install can-isotp==1.9.0` gedowngradet, damit ich das ganze zum laufen bringen konnte. Das wollte ich eigentlich so laufen lassen. Also mit anderen Worten, ich hab einfach die Anleitung befolgt, bevor die Änderungen eingeflossen sind.

Aber kann es sein, dass mich die Timeouts treffen, die Du (@MyHomeMyData) in dieser Antwort meintest? https://github.com/open3e/open3e/discussions/59#discussioncomment-8146938

Wie würde ich das System denn dann aktualisieren? Oder soll ich dem Raspi einfach ein neues OS (Bookworm) verpassen und die open3e Anleitung nochmal durchgehen?

**Comment by MyHomeMyData** [comment]
@timogru Man muss Prioritäten setzen 😄 

Die alte Version läuft bei vielen Anwendern stabil und die beiden Fehlermeldungen deuten nicht auf ein Kernel-Timeout Problem hin. Die Meldung von 26.1. ist ein echter Timeout, d.h. es wurde eine Anfrage korrekt auf den Bus gesendet, es kommt aber keine Antwort. Der Fehler vom 20.1. deutet auch auf ein Problem mit dem CAN-Bus hin.

Gib mal folgenden Befehl ein und poste das Ergebnis:
```
ip -details -statistics link show can0
```
Sollten bei "errors", "dropped" usw. Zahlen deutlich größer 0 stehen, gibt es wahrscheinlich ein Problem mit dem CAN Bus. Dann solltest Du Verkabelung, Masseverbindung und Terminierung überprüfen.

**Reply by timogru** [reply]
Puhh! Ganz am Anfang der Installation hatte ich eine Menge Errors und Dropped. Aber mit dem aktuellen Setup schaut es doch ganz gut aus, oder? Also liegt es vielleicht doch an der gedowngradeten Version? 

```
 $ ip -details -statistics link show can0
5: can0: <NOARP,UP,LOWER_UP,ECHO> mtu 16 qdisc pfifo_fast state UP mode DEFAULT group default qlen 10
    link/can  promiscuity 0  allmulti 0 minmtu 0 maxmtu 0
    can state ERROR-ACTIVE restart-ms 100
          bitrate 250000 sample-point 0.875
          tq 250 prop-seg 6 phase-seg1 7 phase-seg2 2 sjw 1 brp 12
          gs_usb: tseg1 1..16 tseg2 1..8 sjw 1..4 brp 1..1024 brp_inc 1
          clock 48000000
          re-started bus-errors arbit-lost error-warn error-pass bus-off
          0          0          0          0          0          0         numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535 tso_max_size 65536 tso_max_segs 65535 gro_max_size 65536 parentbus usb parentdev 1-1.3:1.0
    RX:  bytes  packets errors dropped  missed   mcast
     114624422 16218069      0       0       0       0
    TX:  bytes  packets errors dropped carrier collsns
       1293880   161735      0       0       0       0
```

**Comment by MyHomeMyData** [comment]
Ja, das sieht sehr gut aus. Dann ist es vermutlich das beste, Du setzt mit Bookworm neu auf. Du kannst den master Branch verwenden, der ist aktuell.

Falls Du mit der Installation nicht klar kommst, einfach nochmal melden. Und natürlich würde mich interessieren, ob das System dann stabil läuft.

**Comment by timogru** [comment]
So, ich hab es installiert und es läuft wieder. Wenn es in fünf Tagen noch Daten liefert, dann werte ich das als Erfolg.
Aber die Inbetriebnahme ging diesmal ohne Probleme, daher hab ich ein gutes Gefühl.

Nochmals Danke für die tolle Hilfe!!

**Reply by timogru** [reply]
Meine finale Rückmeldung: Alles läuft wie geschmiert! Danke für die Unterstützung!
